xapian-core/include/xapian/weight.h

```diff
@@ -409,6 +409,11 @@ class XAPIAN_VISIBILITY_DEFAULT TfIdfWeight : public Weight {
     /// The factor to multiply with the weight.
     double factor;
 
+    /// Parameters slope and delta in the Piv+ normalization weighting formula.
+    /// Default values: 0.2 and 1.0 respectively.
+    double param_slope = 0.2;
+    double param_delta = 1.0;
+
     TfIdfWeight * clone() const;
```

Comment:
As things stand, you initialise the new parameters to non-zero values in the class definition:

```
    double param_slope = 0.2;
    double param_delta = 1.0;
```

So actually I could write: `TfIdfWeight weight("PPP");` and it would work, except that the stats aren't requested correctly in this case.

> Making a way for initialisation of `param_slope` and `param_delta` requires changing the parameter list of this constructor which would break existing tests using this constructor e.g. `TfIdfWeight("ntn")`.

No, you're already initialising them (as above), though as I already noted I don't think you should do it like that, as that's a C++11 feature, which we don't want in the public headers currently - we need to instead initialise in the constructor definition (in `tfidfweight.cc`) like so:

```
TfIdfWeight::TfIdfWeight(const std::string &normals)
    : normalizations(normals), param_slope(0.2), param_slope(1.0)
{
```

The default constructor doesn't need these parameters initialising, as the normalisations is fixed as `"ntn"` in that case.

Are those default parameter values actually sensible choices for a typical user?

If not, then I think we need to reject attempts to use the `P` normalisations without explicitly specifying these parameters.

If they are, then we need to request the stats correctly, and based on whether the normalisation(s) which used them are `P`, not whether the parameters are set (because they are set by default).